core/state/state_witness.go

```diff
@@ -0,0 +1,390 @@
+package state
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"math/big"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// A Witness encompasses a block and all state necessary to compute the
+// post-state root.
+type Witness struct {
+	Block       *types.Block
+	blockHashes map[uint64]common.Hash
+	codes       map[common.Hash][]byte
+	root        common.Hash
+	lists       map[common.Hash]map[string][]byte
+}
+
+// BlockHash returns the block hash corresponding to a historical block (num is assumed to be in the range 1-256)
+func (w *Witness) BlockHash(num uint64) common.Hash {
+	return w.blockHashes[num]
+}
+
+// Root returns the post-state root of the witness if it has been computed or 0x00..0 if not.
+func (w *Witness) Root() common.Hash {
+	return w.root
+}
+
+// rlpWitness is the encoding structure for a Witness
+type rlpWitness struct {
+	EncBlock    []byte
+	Root        common.Hash
+	Owners      []common.Hash
+	AllPaths    [][]string
+	AllNodes    [][][]byte
+	BlockNums   []uint64
+	BlockHashes []common.Hash
+	Codes       [][]byte
+	CodeHashes  []common.Hash
```

Comment:
can CodeHashes be omitted since they can(and should?) be recalculated from Codes?

Reply:
One case where the code wouldn't be necessary to provide or recompute is when an contract account is a target of `EXTCODEHASH` but is never executed or accessed otherwise in the block.

However, I think that for simplicity of the witness format, we should only provide code and calculate/verify the hash as part of the verification.

Thanks for pointing this out!